### PR TITLE
Add IPFS machine backend

### PR DESCRIPTION
**Pull request by geigerzaehler**
* Add a `MachineBackend` implementation based on IPFS
* Extend `test/machine-backends.rad` to test IPFS backend against real IPFS daemon
* Add commands `rad-ipfs`, `rad-ipfs-daemon`, and `rad-ipfs-machine-create`

Follow-up

* Replace `Wreq` with `servant-client` and `ipfs-api` from   https://github.com/oscoin/ipfs/tree/master/ipfs-api
* Move aome code to `ipfs-api`

### CODE REVIEW

File: package.yaml [hunk 1]
```diff
@@ -30,12 +30,16 @@ library:
   - generics-eot
   - haskeline
   - http-client
+  - ipld-cid
   - megaparsec
   - mtl
+  - lens
```
Comment (cloudhead) [inline package.yaml]: Are we using Prisms? In the spirit of smaller build times, might be worth using micro-lens here, if we aren't using Prisms etc.

Comment (cloudhead) [reply package.yaml]: Though I just see you're using `wreq`, which I think depends on `lens` anyway, so maybe ignore..

File: src/Radicle/Internal/MachineBackend/Ipfs.hs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    toJSON MachineEntry{..} =
+        let code = T.intercalate "\n" $ map renderCompactPretty entryExpressions
+        in Aeson.object
+            [ "expression" .= code
+            , "previous" .= ipldLink entryPrevious
+            ]
+
+-- | Given a CID @"abc...def"@ it returns a IPLD link JSON object
+-- @{"/": "abc...def"}@.
+ipldLink :: CID -> Aeson.Value
```
Comment (kim) [inline src/Radicle/Internal/MachineBackend/Ipfs.hs]: This would be nice to have in [oscoin/ipfs](https://github.com/oscoin/ipfs) somewhere

Comment (geigerzaehler) [reply src/Radicle/Internal/MachineBackend/Ipfs.hs]: Yes indeed. There is a couple of things we should backport, like the response types.

File: package.yaml (hunk 1 above)
Comment (geigerzaehler) [reply package.yaml]: I’m switching to `microlens` in our code. Unfortunately `wreq` still has the `lens` dependency. We will get of `wreq` soon.

File: snapshot.yaml [hunk 3]
```diff
@@ -9,6 +9,19 @@ name: radicle-deps-13.2
...
+- serialise-0.2.1.0
+- git: https://github.com/oscoin/ipfs.git
+  commit: 19b25fa4d0003b5c2027c8cc1b8a3c36ac2f9f60
+  subdirs:
+  - ipld-cid
+  - binary-varint
+  - multibase
+  - multihash-cryptonite
+
+flags:
```
Comment (kim) [inline snapshot.yaml]: Dang. bump oscoin/ipfs#7

File: src/Radicle/Internal/MachineBackend/Ipfs.hs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    -- | JSON response from the IPFS Api cannot be parsed. First
+    -- argument is the request path, second argument the JSON parsing
+    -- error
+    | IpfsExceptionInvalidResponse Text Text
+    -- | The IPFS daemon is not running.
+    | IpfsExceptionNoDaemon
+    deriving (Show, Eq)
+
+instance Exception IpfsException where
+    displayException (IpfsException msg) = "ipfs: " <> toS msg
```
Comment (kim) [inline src/Radicle/Internal/MachineBackend/Ipfs.hs]: TIL `displayException` is a thing

File: src/Radicle/Internal/MachineBackend/Ipfs.hs [hunk 5]
```diff
@@ -0,0 +1,318 @@
+-- | This module implements an IPFS machine backend.
+--
+-- Radicle Machines are identified by IPNS IDs. The IPNS ID of a
+-- machine points to a linked listed of expressions encoded as content
```
Comment (jameshaydon) [inline src/Radicle/Internal/MachineBackend/Ipfs.hs]:
```suggestion
-- machine points to a linked list of expressions encoded as content
```

**Pull request merged**